ge0/geo_url_parser.cpp

```diff
@@ -152,6 +155,136 @@ int LatLonParser::GetCoordinatesPriority(string const & token)
   return -1;
 }
 
+std::string const kLatLon = R"(([+-]?\d+(?:\.\d+)?), *([+-]?\d+(?:\.\d+)?)(:?, *([+-]?\d+(?:\.\d+)?))?)";
+
+GeoParser::GeoParser()
+  : m_latlonRe(kLatLon), m_zoomRe(kFloatIntScale)
+{
+}
+
+bool GeoParser::Parse(std::string const & raw, GeoURLInfo & info) const
+{
+  info.Reset();
+
+  /*
+   * References:
+   * - https://datatracker.ietf.org/doc/html/rfc5870
+   * - https://developer.android.com/guide/components/intents-common#Maps
+   * - https://developers.google.com/maps/documentation/urls/android-intents
+   */
+
+  /*
+   * Check that URI starts with geo:
+   */
+  if (raw.size() < 4 || raw.substr(0, 4) != "geo:")
+    return false;
+
+  /*
+   * Check for trailing `(label)` which is not RFC3986-compliant (thanks, Google).
+   */
+  size_t end = string::npos;
+  if (raw.size() > 2 && raw[raw.size() - 1] == ')' && string::npos != (end = raw.rfind('(', raw.size() - 1)))
+  {
+    // head (label)
+    //      ^end
+    info.m_label = url::UrlDecode(raw.substr(end + 1, raw.size() - end - 2));
+    // Remove any whitespace between `head` and `(`.
+    end--;
+    while (end > 0 && (raw[end] == ' ' || raw[end] == '+'))
+      end--;
+  }
+
+  url::Url url(end == string::npos ? raw : raw.substr(0, end + 1));
+  if (!url.IsValid())
+    return false;
+  ASSERT_EQUAL(url.GetScheme(), "geo", ());
+
+  /*
+   * Parse coordinates before ';' character
+   */
+  std::string coordinates = url.GetHost().substr(0, url.GetHost().find(';'));
```

Comment:
Note that if ';' is not in GetHost(), coordinates string will be equal to url.GetHost(), not empty. Is that what you mean?

Reply:
Yes, I want a string before ";" or entire Host instead.